File: client/crypto.py
#!/usr/bin/python

# Copyright 2012 Patrick Moor <patrick@moor.ws>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import base64
import M2Crypto
from M2Crypto import EVP
from M2Crypto.Rand import rand_bytes

__author__ = 'pmoor'

def createEncryption(secret_key, value):
  iv = rand_bytes(16)
  cipher = EVP.Cipher(alg="aes_256_cbc", key=secret_key, iv=iv, op=M2Crypto.encrypt)
  ciphertext = cipher.update(value)
  ciphertext += cipher.final()
  hmac = EVP.hmac(secret_key, iv + value, algo="sha256")

  return {
    "version": 0,
    "algorithm": "AES_256_CBC_HMAC_SHA256",
    "iv": base64.standard_b64encode(iv),
    "ciphertext": base64.standard_b64encode(ciphertext),
    "signature": base64.standard_b64encode(hmac)
  }

def newRandomSecretKey():
  return rand_bytes(256 / 8)


def createEncryptionWithPassphrase(secret_key, passphrase):
  iv_and_salt = rand_bytes(128 / 8)
  password_derived_key = EVP.pbkdf2(passphrase, iv_and_salt, 1000, 256 / 8)

  cipher = EVP.Cipher(alg="aes_256_cbc", key=password_derived_key, iv=iv_and_salt, op=M2Crypto.encrypt)
  encrypted_key = cipher.update(secret_key)
  encrypted_key += cipher.final()
  hmac = EVP.hmac(password_derived_key, iv_and_salt + secret_key, algo="sha256")

  return {
    "version": 0,
    "algorithm": "PBKDF2_SHA1_AES_256_CBC_HMAC_SHA256",
    "iv": base64.standard_b64encode(iv_and_salt),
    "ciphertext": base64.standard_b64encode(encrypted_key),
    "signature": base64.standard_b64encode(hmac)
  }

def extractSecretKey(key_resource, passphrase):
  for encryption in key_resource["encryptions"]:
    assert encryption["algorithm"] == "PBKDF2_SHA1_AES_256_CBC_HMAC_SHA256"
    iv = base64.standard_b64decode(encryption["iv"])
    ciphertext = base64.standard_b64decode(encryption["ciphertext"])
    signature = base64.standard_b64decode(encryption["signature"])

    password_derived_key = EVP.pbkdf2(passphrase, iv, 1000, 256 / 8)

    try:
      cipher = EVP.Cipher(alg="aes_256_cbc", key=password_derived_key, iv=iv, op=M2Crypto.decrypt)
      secret_key = cipher.update(ciphertext)
      secret_key += cipher.final()
    except EVP.EVPError:
      # try other encryptions
      continue

    hmac = EVP.hmac(password_derived_key, iv + secret_key, algo="sha256")
    if hmac == signature:
      return secret_key


def decryptSecret(encryption, secret_key):
  assert encryption["algorithm"] == "AES_256_CBC_HMAC_SHA256"
  iv = base64.standard_b64decode(encryption["iv"])
  ciphertext = base64.standard_b64decode(encryption["ciphertext"])
  signature = base64.standard_b64decode(encryption["signature"])

  cipher = EVP.Cipher(alg="aes_256_cbc", key=secret_key, iv=iv, op=M2Crypto.decrypt)
  value = cipher.update(ciphertext)
  value += cipher.final()

  hmac = EVP.hmac(secret_key, iv + value, algo="sha256")
  if hmac != signature:
    raise Exception("invalid signature")
  return value

File: src/main/webapp/js/crypto.js
/*
  Copyright 2012 Patrick Moor <patrick@moor.ws>

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

// AES block size: 128bit (16 bytes, 4 words)
// AES IV size: 128bit
// AES key size: 256bit (32 bytes, 8 words)

goog.provide('swissvault.crypto');

swissvault.crypto.encrypt = function(key, cleartextUtf8) {
  var cleartext = Crypto.charenc.UTF8.stringToBytes(cleartextUtf8);

  var iv = Crypto.util.randomBytes(16);
  var signature = Crypto.HMAC(Crypto.SHA256, iv.concat(cleartext), key, {asBytes: true});

  // this messes with cleartext
  var ciphertext = Crypto.AES.encrypt(
      cleartext, key, { mode: new Crypto.mode.CBC(Crypto.pad.pkcs7), iv: iv, asBytes: true });

  return {
    'version': 0,
    'algorithm': 'AES_256_CBC_HMAC_SHA256',
    'iv': Crypto.util.bytesToBase64(iv),
    'ciphertext': Crypto.util.bytesToBase64(ciphertext),
    'signature': Crypto.util.bytesToBase64(signature)
  }
};

swissvault.crypto.decrypt = function(key, encryption, asBytes) {
  var iv = Crypto.util.base64ToBytes(encryption.iv);
  var ciphertext = Crypto.util.base64ToBytes(encryption.ciphertext);
  var signature = encryption.signature;

  var cleartext = Crypto.AES.decrypt(
      ciphertext, key, { mode: new Crypto.mode.CBC(Crypto.pad.pkcs7), iv: iv, asBytes: true });

  var expectedSignature = Crypto.util.bytesToBase64(
      Crypto.HMAC(Crypto.SHA256, iv.concat(cleartext), key, {asBytes: true}));
  if (expectedSignature != signature) {
    throw new Error("signature does not match");
  }

  return asBytes ? cleartext : Crypto.charenc.UTF8.bytesToString(cleartext);
};

swissvault.crypto.newEncryption = function(passphraseUtf8) {
  var passphrase = Crypto.charenc.UTF8.stringToBytes(passphraseUtf8);

  var secret_key_part_1 = Crypto.util.randomBytes(32);
  var secret_key_part_2 = Crypto.SHA256(passphrase, {asBytes: true});
  var secret_key = new Array(32);
  for (var i = 0; i < 32; i++) {
    secret_key[i] = secret_key_part_1[i] ^ secret_key_part_2[i];
  }

  return swissvault.crypto.encryptWithPassphrase(passphraseUtf8, secret_key);
};

swissvault.crypto.encryptWithPassphrase = function(passphraseUtf8, secret_key) {
  var passphrase = Crypto.charenc.UTF8.stringToBytes(passphraseUtf8);

  var iv = Crypto.util.randomBytes(16);
  var key = Crypto.PBKDF2(passphrase, iv, 32, {iterations: 1000, asBytes: true});
  
  var signature = Crypto.HMAC(Crypto.SHA256, iv.concat(secret_key), key, {asBytes: true});
  var ciphertext = Crypto.AES.encrypt(
      secret_key.slice(0), key, { mode: new Crypto.mode.CBC(Crypto.pad.pkcs7), iv: iv, asBytes: true });

  return {
    'version': 0,
    'algorithm': 'PBKDF2_SHA1_AES_256_CBC_HMAC_SHA256',
    'iv': Crypto.util.bytesToBase64(iv),
    'ciphertext': Crypto.util.bytesToBase64(ciphertext),
    'signature': Crypto.util.bytesToBase64(signature)
  }
};

swissvault.crypto.tryEncryptionDecryption = function(encryption, passphraseUtf8) {
  var passphrase = Crypto.charenc.UTF8.stringToBytes(passphraseUtf8);
  var salt = Crypto.util.base64ToBytes(encryption.iv);
  var key = Crypto.PBKDF2(passphrase, salt, 32, {iterations: 1000, asBytes: true});
  try {
    return swissvault.crypto.decrypt(key, encryption, true);
  } catch(e) {
    return null;
  }
};